libr/bin/bin.c

```diff
@@ -349,73 +347,56 @@ R_API bool r_bin_open_io(RBin *bin, RBinOptions *opt) {
 	if (opt->sz >= UT32_MAX) {
 		opt->sz = 1024 * 32;
 	}
-
+	RBuffer *buf = r_buf_new_with_io (&bin->iob, tfd);
+	if (!buf) {
+		return false;
+	}
 	bin->file = fname;
 	opt->sz = R_MIN (file_sz, opt->sz);
-	if (!r_list_length (bin->binfiles)) {
-		if (is_debugger) {
-			//use the temporal RIODesc to read the content of the file instead
-			//from the memory
-			if (tfd >= 0) {
-				buf_bytes = calloc (1, opt->sz + 1);
-				if (buf_bytes) {
-					iob->fd_read_at (io, tfd, 0, buf_bytes, opt->sz);
-				}
-				// iob->fd_close (io, tfd);
-			}
-		}
-	}
-	// this thing works for 2GB ELF core from vbox
-	if (!buf_bytes) {
-		const unsigned int asz = opt->sz? (unsigned int)opt->sz: 1;
-		buf_bytes = calloc (1, asz);
-		if (!buf_bytes) {
-			eprintf ("Cannot allocate %u bytes.\n", asz);
-			return false;
-		}
-		ut64 seekaddr = is_debugger? opt->baseaddr: opt->loadaddr;
-		if (!iob->fd_read_at (io, opt->fd, seekaddr, buf_bytes, asz)) {
-			opt->sz = 0LL;
+	ut64 seekaddr = is_debugger? opt->baseaddr: opt->loadaddr;
+	if (seekaddr > 0 && seekaddr != UT64_MAX) {
+		// slice buffer if necessary
+		RBuffer *nb = r_buf_new_slice (buf, seekaddr, opt->sz);
```

Comment:
why do we need to slice? i think we need it only when working on fat binaries, where you want to consider a binary that starts in the middle of another file.